File: javascript/hamming/hamming.js
function Hamming() {
}

Hamming.prototype.compute = function (strand1, strand2) {
  if (strand1.length !== strand2.length) {
    throw new Error('DNA strands must be of equal length.');
  }

  var differentNucleotides = strand1
    .split('')
    .filter(function (nucleotide, i) {
      return nucleotide !== strand2[i];
    });

  return differentNucleotides.length;
}

module.exports = new Hamming();

File: elixir/hamming/hamming.exs
defmodule Hamming do
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(strand1, strand2) do
    hamming(strand1, strand2, 0)
  end

  defp hamming([], [], distance), do: { :ok, distance }
  defp hamming([], strand2, _), do: { :error, "Lists must be the same length"}
  defp hamming(strand1, [], _), do: { :error, "Lists must be the same length"}
  defp hamming([ a | strand1 ], [ a | strand2 ], distance), do:
    hamming(strand1, strand2, distance)
  defp hamming([ a | strand1 ], [ b | strand2 ], distance), do:
    hamming(strand1, strand2, distance + 1)

end
